aignostic/aignostic/dataset/validate_dataset_api.py

```
@@ -1,65 +1,63 @@
 
-    if any(len(dataset.column_names) != len(row) for row in dataset.rows):
-        raise ValidationError("Number of column names not equal to the number of elements in one or more rows")
+    # Validate inner list consistency in a single loop
+    feature_length = len(features[0]) if features else 0
+    label_length = len(labels[0]) if labels else 0
 
-    return dataset
+    for f_row, l_row in zip(features, labels):
+        if len(f_row) != feature_length:
+            raise ValueError("All feature rows must have the same number of elements.")
+        if len(l_row) != label_length:
+            raise ValueError("All label rows must have the same number of elements.")
 
+    return data
 
-@app.get('/validate-dataset')
-async def validate_dataset(url: str = Query(..., description="Dataset URL")):
-    """
-    Validate a dataset URL and return the columns and number of rows.
+
+@app.post("/fetch-data")
+def fetch_dataset(request: FetchDatasetRequest) -> ModelInput:
     """
-    if not is_valid_url(url):
-        raise HTTPException(status_code=400, detail="Invalid URL")
+    Validate a dataset URL and parse the data to be used in the model.
 
+    Params:
+    - request : FetchDatasetRequest - Pydantic model for the request
+    """
     try:
-        data = fetch_data(url)
-        dataset = parse_dataset(data)
-        return JSONResponse(content={
-            "message": "Data successfully parsed into DataFrame",
-            "columns": dataset.column_names,
-            "rows": len(dataset.rows)
-        })
+        headers = {"Authorization": f"Bearer {request.dataset_api_key}"} if request.dataset_api_key else {}
+        response = requests.get(request.dataset_url, headers=headers)
+
+        response.raise_for_status()
 
-    # TODO: Add logging to exception handling
+        data = response.json()
+        return validate_dataset_format(data)
     except requests.exceptions.RequestException as e:
```

I don't think we should rewrite any of the exception logic (or catch it) -> for me it makes sense to make 4XX error handling the responsibility of the client and we can propagate the error straight to the frontend without modifying it. This will likely make the errors easier for the client to understand as it's their own error message.